proposals/NNNN-unlock-existentials-with-fixed-associated-types.md

```diff
@@ -0,0 +1,249 @@
+
+```swift
+protocol User: Identifiable 
+    where ID == String {
+    
+    var username: String { get }
+    var displayName: String { get }
+}
+
+extension User {
+    var id: String {
+        username
+    }
+}
+```
+
+Many would rightfully assume that `User` could be used as a Type, since we specify that the `ID` asscoated type inherited from [`Identifiable`](https://github.com/apple/swift-evolution/blob/master/proposals/0261-identifiable.md) is `String`. 
+
+```swift
+let myUser: User
+// ❌ `User` is a PUT
+```
+
+This is a great inconvenience with not so elegant workarounds:
+
+```swift
+protocol AnyUser { ... } 
+// No `Identifiable` conformance
+
+let myUser: AnyUser
+// ✅
+
+
+protocol User: AnyUser, Identifiable 
+    where ID == String { 
+    
+    var id: String {
+        username
+    }
+}
+
+let myOtherUser: User 
+// ❌ Still an error as expected
+```
+
+These workarounds, besides producing boilerplate and confusing code, also produce confusing API designs leaving clients to wonder which of the two protocols should be used. In other cases, API authors may decide that adding `Identifiable` isn’t worth the added complexity and leave the `User` protocol with no such conformance whatsoever. 
+
+As a result, a straightforward conformance to simple - yet quite useful - protocols such as `Identifiable` and others becomes a hassle to implement.
+
+
+## Proposed solution
+
+We propose to simply stop consider User-like protocols PUTs and instead treat them as regular protocols, enabling use as Types. Thus, making more natural code possible - which fits a goal of Swift of building ["expressive and elegant APIs"](https://forums.swift.org/t/on-the-road-to-swift-6/32862). Furthermore, library authors could start adding useful protocol conformances to their protocols - a task that's currently prohibitively complex.
+
+## Detailed design
+
+### What would be a Regular Protocol?
+
+What would protocols would be able to be used as types (✅) and what wouldn’t (❌)?
+
+1. ✅ Simple Case (like with `User`)
+```swift
+protocol User: Identifiable
+    where ID == String { ... }
+```
+
+2. ❌ Multiple Associated Types; only Some Specified 
+```swift
+protocol PUT { 
+    associatedtype A
+    associatedtype B
+}
+
+protocol AlsoPUT: PUT
+    where A == String { ... } 
+// `B` is unknown; therefore it’s a PUT
+```
+
+3. ✅ Multiple Associated Types; All Specified
+```swift
+protocol UsableAsAType: PUT
+    where A == String, B == Int { ... }
+```
+
+4. ✅ Composite Type 
+```swift
+typealias Bar = UsableAsAType & User
+```
+
+5. ❌ Protocols Referencing `Self` (currently erroneous)
+```swift
+protocol PUT: Equatable
+    where Self == String { ... }
+❌ This requirement makes `Self` non-generic
+```
+> **_NOTE:_** An `Equatable` conforming protocol will still be PUT, because if `Self` is to be constrained then the protocol loses its generic meaning.
+
+6. ✅ Constraining with the Protocol Existential Itself
```

Comment:
Some of these examples are redundant or irrelevant. `6.` is equivalent to `3.` and `1.`, `5.` shows an *invalid* protocol. Here's an initial list of cases that are worth an example:
* A protocol with all of its associated types fixed (let the number of associated types be 2)
* A protocol like the previous one, but with a requirement referencing `Self` in function parameter position (we emphasize that this still makes the protocol unsupported)
* A protocol composition where both protocols have their associated types fixed
* A subclass existential, where protocol(s) have their associated types fixed and the class does **not** conform to anything.

Reply:
Also, I think it's important to have a case where the protocol has partially fixed associated types:

```swift
protocol AB {
    associatedtype A
    associatedtype B
}

protocol FixedA: AB 
    where A == String {}

let foo: FixedA
// ❌
```

Reply:
> What exactly do you mean by your fourth item on the list, could you give an example? 

```Protocol & Class```

> Also, I think it's important to have a case where the protocol has partially fixed associated types

We should be explaining that a protocol with associated type requirements will only be supported as a value type if

* *all* of its associated types are known to be concrete, and
* there are no `Self` references in requirements that are *already* preventing it from being supported

IMO the first bullet makes the outcome of your example pretty clear.